### PR TITLE
Adding Windows Services, Registry Core Ext from PT Run

### PR DESCRIPTION
**Pull request by joadoumie**
First pass at adding the equivalent functionality of Windows Services from PT Run.

### CODE REVIEW

**Comment by joadoumie** [comment]
This PR includes first passes for both the Windows Services Plugin & the Windows Registry Plugin

File: src/modules/cmdpal/Exts/Microsoft.CmdPal.Ext.WindowsServices/Properties/Resources.resx [hunk 1]
```diff
@@ -0,0 +1,210 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (zadjii-msft) [inline src/modules/cmdpal/Exts/Microsoft.CmdPal.Ext.WindowsServices/Properties/Resources.resx:1]: 👀 we actually didn't really have any localized components before this. Glad that it just worked?

File: src/modules/cmdpal/Exts/Microsoft.CmdPal.Ext.WindowsServices/Properties/Resources.Designer.cs [hunk 2]
```diff
@@ -0,0 +1,333 @@
+﻿//------------------------------------------------------------------------------
```
Comment (zadjii-msft) [inline src/modules/cmdpal/Exts/Microsoft.CmdPal.Ext.WindowsServices/Properties/Resources.Designer.cs:1]: @crutkas I need a C# adult - this looks like the kind of file that should be .gitignored?

File: src/modules/cmdpal/Exts/Microsoft.CmdPal.Ext.WindowsServices/Pages/ServicesListPage.cs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+{
+    public ServicesListPage()
+    {
+        Icon = new(string.Empty);
+        Name = "Windows Services";
+    }
+
+    public override ISection[] GetItems(string query)
+    {
+        ListItem[] items = ServiceHelper.Search(query).ToArray();
```
Comment (zadjii-msft) [inline src/modules/cmdpal/Exts/Microsoft.CmdPal.Ext.WindowsServices/Pages/ServicesListPage.cs:32]: 📝 we could totally cache this result in the future, but good enough for v1

File: src/modules/cmdpal/Exts/Microsoft.CmdPal.Ext.WindowsServices/Helpers/ServiceHelper.cs [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+                Subtitle = ServiceHelper.GetResultSubTitle(s),
+                MoreCommands = moreCommands,
+
+                // ToolTipData = new ToolTipData(serviceResult.DisplayName, serviceResult.ServiceName),
+                // IcoPath = icoPath,
+            };
+        });
+    }
+
+    // TODO: IPublicAPI contextAPI isn't used anymore, but we need equivalent ways to show notifications and status
```
Comment (zadjii-msft) [inline src/modules/cmdpal/Exts/Microsoft.CmdPal.Ext.WindowsServices/Helpers/ServiceHelper.cs:81]: Something like https://github.com/zadjii-msft/PowerToys/issues/118?

File: src/modules/cmdpal/Exts/Microsoft.CmdPal.Ext.WindowsServices/Helpers/ServiceHelper.cs [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+            }
+
+            return new ListItem(serviceAction)
+            {
+                Title = s.DisplayName,
+                Subtitle = ServiceHelper.GetResultSubTitle(s),
+                MoreCommands = moreCommands,
+
+                // ToolTipData = new ToolTipData(serviceResult.DisplayName, serviceResult.ServiceName),
+                // IcoPath = icoPath,
```
Comment (zadjii-msft) [inline src/modules/cmdpal/Exts/Microsoft.CmdPal.Ext.WindowsServices/Helpers/ServiceHelper.cs:76]: maybe leave a comment linking to #78

File: src/modules/cmdpal/Exts/Microsoft.CmdPal.Ext.WindowsServices/Commands/ServiceCommand.cs [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    private readonly Action _action;
+
+    internal ServiceCommand(ServiceResult serviceResult, Action action)
+    {
+        _serviceResult = serviceResult;
+        _action = action;
+        Name = action.ToString();
+        if (serviceResult.IsRunning)
+        {
+            Icon = new("\xE71A");
```
Comment (zadjii-msft) [inline src/modules/cmdpal/Exts/Microsoft.CmdPal.Ext.WindowsServices/Commands/ServiceCommand.cs:31]: nit: these could use a comment for what icon they are

File: src/modules/cmdpal/Exts/Microsoft.CmdPal.Ext.Registry/Helpers/ResultHelper.cs [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+            else if (entry.Key is null && !(entry.Exception is null))
+            {
+                // on error (e.g access denied)
+                // TODO: Investigate query text display (i.e placeholder text?) --> result.QueryTextDisplay = entry.KeyPath;
+                result.Subtitle = GetTruncatedText(entry.Exception.Message, MaxTextLength.MaximumSubTitleLengthWithTwoSymbols, TruncateSide.OnlyFromRight);
+                result.Title = GetTruncatedText(entry.KeyPath, MaxTextLength.MaximumTitleLengthWithTwoSymbols);
+            }
+            else
+            {
+                // TODO: Investigate query text display (i.e placeholder text?) --> result.QueryTextDisplay = entry.KeyPath;
```
Comment (zadjii-msft) [inline src/modules/cmdpal/Exts/Microsoft.CmdPal.Ext.Registry/Helpers/ResultHelper.cs:58]: link to #121 (and above too)

File: src/modules/cmdpal/Exts/Microsoft.CmdPal.Ext.Registry/Helpers/ResultHelper.cs [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+                result.Title = GetTruncatedText(entry.KeyPath, MaxTextLength.MaximumTitleLengthWithTwoSymbols);
+            }
+
+            // result.ContextData = entry;
+            // TODO: Investigate tool tips, result.ToolTipData = new ToolTipData(Resources.RegistryKey, $"{Resources.KeyName} {result.Title}");
```
Comment (zadjii-msft) [inline src/modules/cmdpal/Exts/Microsoft.CmdPal.Ext.Registry/Helpers/ResultHelper.cs:63]: link this bare TODO to an issue

File: src/modules/cmdpal/Exts/Microsoft.CmdPal.Ext.Registry/Commands/OpenKeyInEditorCommand.cs [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    internal static bool TryToOpenInRegistryEditor(in RegistryEntry entry)
+    {
+        try
+        {
+            RegistryHelper.OpenRegistryKey(entry.Key?.Name ?? entry.KeyPath);
+            return true;
+        }
+        catch (System.ComponentModel.Win32Exception)
+        {
+            // TODO: We need a convenient way to show errors to the user
```
Comment (zadjii-msft) [inline src/modules/cmdpal/Exts/Microsoft.CmdPal.Ext.Registry/Commands/OpenKeyInEditorCommand.cs:48]: link to #118

File: src/modules/cmdpal/Exts/Microsoft.CmdPal.Ext.Registry/Commands/CopyValueNameCommand.cs [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+    internal CopyValueNameCommand(RegistryEntry entry)
+    {
+        Name = Resources.CopyValueName;
+        Icon = new("\xE8C8");
+        _entry = entry;
+    }
+
+    private static bool TryToCopyToClipBoard(in string text)
+    {
+        // TODO: Have this actually use the clipboard helper
```
Comment (zadjii-msft) [inline src/modules/cmdpal/Exts/Microsoft.CmdPal.Ext.Registry/Commands/CopyValueNameCommand.cs:35]: 🛑: You should be able to do this now

File: src/modules/cmdpal/Exts/Microsoft.CmdPal.Ext.Registry/Commands/CopyValueDataCommand.cs [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+    internal CopyValueDataCommand(RegistryEntry entry)
+    {
+        Name = Resources.CopyValueData;
+        Icon = new("\xF413");
+        _entry = entry;
+    }
+
+    private static bool TryToCopyToClipBoard(in string text)
+    {
+        // TODO: Have this actually use the clipboard helper
```
Comment (zadjii-msft) [inline src/modules/cmdpal/Exts/Microsoft.CmdPal.Ext.Registry/Commands/CopyValueDataCommand.cs:35]: 🛑: You should be able to do this now

File: src/modules/cmdpal/Exts/Microsoft.CmdPal.Ext.Registry/Commands/OpenKeyInEditorCommand.cs [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+        Name = Resources.OpenKeyInRegistryEditor;
+        Icon = new("\xE8A7");
+        _entry = entry;
+    }
+
+    private static bool TryToCopyToClipBoard(in string text)
+    {
+        // TODO: Have this actually use the clipboard helper
+        return true;
+    }
```
Comment (zadjii-msft) [inline src/modules/cmdpal/Exts/Microsoft.CmdPal.Ext.Registry/Commands/OpenKeyInEditorCommand.cs:37]: we don't need this in this class, do we?

File: src/modules/cmdpal/Exts/Microsoft.CmdPal.Ext.Registry/Commands/CopyValueNameCommand.cs [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+using Windows.ApplicationModel.DataTransfer;
+using Windows.UI;
+
+namespace Microsoft.CmdPal.Ext.Registry.Commands;
+
+internal sealed partial class CopyValueNameCommand : InvokableCommand
+{
+    private readonly RegistryEntry _entry;
+
+    internal CopyValueNameCommand(RegistryEntry entry)
```
Comment (zadjii-msft) [inline src/modules/cmdpal/Exts/Microsoft.CmdPal.Ext.Registry/Commands/CopyValueNameCommand.cs:26]: you could _probably_ dedupe this with CopyValueDataCommand by just passing in a `bool copyName` to indicate if you want to copy the name or data

Comment (zadjii-msft) [reply src/modules/cmdpal/Exts/Microsoft.CmdPal.Ext.Registry/Commands/CopyValueNameCommand.cs:26]: actually the same with the CopyKey one above that too. Maybe pass in an enum for which of the three you want to copy

**Review by zadjii-msft — CHANGES_REQUESTED** [review]
okay I reviewed the first two commits - services and registry

* I didn't look too hard at the implementation bits that are clearly straight outta PT run. I assume that code's already fine.
* I left a bunch of notes on linking bare `TODO`s - we're in the real game now (not just hackathon), so we should have TODO's in the code reference an actual work item tracking that TODO. Otherwise we just lose them. In terminal we've used `TODO GH #xyz: <comment>`to help keep track of the issue that's tracking that TODO
* I don't know about the resources thing so clint halp pls
  * I'm blocking on this
* we probably want to actually implement the copy thing now
  * i'm soft-blocking on this
* revert the settings search one and the commit after that, and file a stacked PR for settings search 😜

File: src/modules/cmdpal/Exts/Microsoft.CmdPal.Ext.WindowsServices/Helpers/ServiceHelper.cs (hunk 4 above)
Comment (joadoumie) [reply src/modules/cmdpal/Exts/Microsoft.CmdPal.Ext.WindowsServices/Helpers/ServiceHelper.cs:81]: Yeah exactly like that.

File: src/modules/cmdpal/Exts/Microsoft.CmdPal.Ext.WindowsServices/Properties/Resources.resx (hunk 1 above)
Comment (joadoumie) [reply src/modules/cmdpal/Exts/Microsoft.CmdPal.Ext.WindowsServices/Properties/Resources.resx:1]: #124 -- I'm not sure if it is actually localized or not / how that pipeline works. I think it is?

File: src/modules/cmdpal/Exts/Microsoft.CmdPal.Ext.WindowsServices/Pages/ServicesListPage.cs (hunk 3 above)
Comment (joadoumie) [reply src/modules/cmdpal/Exts/Microsoft.CmdPal.Ext.WindowsServices/Pages/ServicesListPage.cs:32]: Hmmmmm good point - let's just do it? It's probably best practice and we can use that as a sample down the line?

HOW should we do it the BEST way (best being best for now that is reasonable, but does some caching)

File: src/modules/cmdpal/Exts/Microsoft.CmdPal.Ext.Registry/Commands/CopyRegistryInfoCommand.cs [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+        }
+
+        _entry = entry;
+    }
+
+    private static bool TryToCopyToClipBoard(in string text)
+    {
+        // TODO: Have this actually use the clipboard helper
+        return true;
+    }
```
Comment (zadjii-msft) [inline src/modules/cmdpal/Exts/Microsoft.CmdPal.Ext.Registry/Commands/CopyRegistryInfoCommand.cs:47]: dead code?

**Comment by crutkas** [comment]
Yes the are included

File: src/modules/cmdpal/Exts/Microsoft.CmdPal.Ext.Registry/Commands/CopyRegistryInfoCommand.cs (hunk 14 above)
Comment (joadoumie) [reply src/modules/cmdpal/Exts/Microsoft.CmdPal.Ext.Registry/Commands/CopyRegistryInfoCommand.cs:47]: Indeed. Good catch. thank you. I DONT know how this is still in here tbh.

**Review by zadjii-msft — APPROVED** [review]

**Comment by joadoumie** [comment]
Ref to #84

**Pull request merged**